ledger/src/blockstore.rs

```diff
@@ -1484,14 +1487,26 @@ impl Blockstore {
         })
     }
 
+    pub fn get_data_shred(&self, slot: Slot, index: u64) -> Result<Option<Vec<u8>>> {
+        Self::fetch_data_shred(slot, index, &self.data_shred_cf)
+    }
+
     pub fn get_data_shreds_for_slot(
         &self,
         slot: Slot,
         start_index: u64,
     ) -> ShredResult<Vec<Shred>> {
+        use crate::shred::SHRED_PAYLOAD_SIZE;
         self.slot_data_iterator(slot, start_index)
             .expect("blockstore couldn't fetch iterator")
-            .map(|data| Shred::new_from_serialized_shred(data.1.to_vec()))
+            .map(|data| {
+                let mut payload = data.1.to_vec();
+                // Only data_header.size bytes stored in the blockstore so
+                // pad the payload out to SHRED_PAYLOAD_SIZE so that the
+                // erasure recovery works properly.
+                payload.resize(cmp::max(payload.len(), SHRED_PAYLOAD_SIZE), 0);
```

Comment:
I'm questioning why we use `cmp::max(payload.len(), SHRED_PAYLOAD_SIZE)` and not just `SHRED_PAYLOAD_SIZE`; we explicitly check for "oversized" shreds in `should_insert_data_shred()` so there `payload.len() > SHRED_PAYLOAD_SIZE` should be impossible
```
    fn should_insert_data_shred(...) -> bool {
    ...
        if shred.payload.len() > SHRED_PAYLOAD_SIZE {
            ...
            return false;
```

The check also occurs in `is_shred_duplicate()` and `fetch_data_shred()`